artifactory/commands/terraform/terraformpublish.go

```diff
@@ -0,0 +1,304 @@
+				e = services.CollectFilesForUpload(*uploadParams, nil, nil, dataHandlerFunc)
+				if e != nil {
+					return e
+				}
+				// SkipDir will not stop the walk, but will jump to the next directory.
+				return filepath.SkipDir
+			}
+			return nil
+		})
+		if err != nil && err != io.EOF {
+			log.Error(err)
+			errorsQueue.AddError(err)
+		}
+
+		// Upload modules
+		for targetPath, archiveData := range toArchive {
+			err := archiveData.GetWriter().Close()
+			if err != nil {
+				log.Error(err)
+				errorsQueue.AddError(err)
+			}
+			// Upload module using upload service
+			serviceManager, err := utils.CreateServiceManager(ts.serverDetails, 0, 0, false)
+			if err != nil {
+				log.Error(err)
+				errorsQueue.AddError(err)
+			}
+			uploadService := services.NewUploadService(serviceManager.Client())
+			uploadService.ArtDetails = serviceManager.GetConfig().GetServiceDetails()
+			uploadService.Threads = serviceManager.GetConfig().GetThreads()
+			producer.AddTaskWithError(uploadService.CreateUploadAsZipFunc(uploadSummary, targetPath, archiveData, errorsQueue), errorsQueue.AddError)
+		}
+	}()
+}
+
+func (ts *TerraformPublishCommand) performTerraformPublishTasks(consumer parallel.Runner, uploadSummary *clientservicesutils.Result) (totalUploaded, totalFailed int) {
+	// Blocking until consuming is finished.
+	consumer.Run()
+	totalUploaded = clientservicesutils.SumIntArray(uploadSummary.SuccessCount)
+	totalUploadAttempted := clientservicesutils.SumIntArray(uploadSummary.TotalCount)
+
+	log.Debug("Uploaded", strconv.Itoa(totalUploaded), "artifacts.")
+	totalFailed = totalUploadAttempted - totalUploaded
+	if totalFailed > 0 {
+		log.Error("Failed uploading", strconv.Itoa(totalFailed), "artifacts.")
+	}
+	return
+}
+
+func (tp *TerraformPublishCommand) uploadParamsForTerraformPublish(moduleName, dirPath string) (*services.UploadParams, error) {
+	uploadParams := services.NewUploadParams()
+	uploadParams.Target = tp.getPublishTarget(moduleName)
+	uploadParams.Pattern = dirPath + "/(*)"
+	uploadParams.TargetPathInArchive = "{1}"
+	uploadParams.Archive = "zip"
+	uploadParams.Recursive = true
+	uploadParams.CommonParams.TargetProps = specutils.NewProperties()
+	uploadParams.CommonParams.Exclusions = append(tp.exclusions, "*.git", "*.DS_Store")
+
+	return &uploadParams, nil
+}
+
+// Module's path in terraform repository : namespace/provider/moduleName/tag.zip
+func (tp *TerraformPublishCommand) getPublishTarget(moduleName string) string {
+	return filepath.ToSlash(filepath.Join(tp.repo, tp.namespace, tp.provider, moduleName, tp.tag+".zip"))
+}
+
+// We identify a Terraform module by the existing of a '.tf' file inside the module directory.
+// isTerraformModule search for '.tf' file inside and returns true it founds at least one.
+func checkIfTerraformModule(path string) (bool, error) {
+	dirname := path + string(filepath.Separator)
+	d, err := os.Open(dirname)
+	if err != nil {
+		return false, err
+	}
+	defer d.Close()
```

Comment:
The error returned here shouldn't be ignored. you should use this pattern for errors inside defer clauses - 
```
	defer func() {
		for _, reader := range temp {
			e := reader.Close()
			if err == nil {
				err = errorutils.CheckError(e)
			}
		}
	}()
``` 
This pattern is very important and should be always used when a function used with defer returns an error.